[SelwynAng] iP

SelwynAng opened the pull request:
# Duke Prince

> "Tracking my tasks, especially in University, is so hard! I wish there is a solution for this..."  ~ Mr [Eyeone Todye ](https://www.wikihow.life/Survive-Your-First-Year-at-University) 😵‍💫 

Do you also face the same issue as this poor dude above? Fret not. **Duke Prince** is here to save the day! 
**Duke Prince** is an awesome task list tracker that helps you create and note such tasks:
- To dos
- Deadlines
- Events


All you need to do is:
1. Download the JAR file provided in the release tab
2. Run the command `java -jar Duke.jar` in your terminal in the folder containing the downloaded JAR file
3. Create your tasks, mark your tasks and get productive!


These are some of the features that the application contains for now:
- [x] Creating, marking, un-marking and deleting of tasks
- [ ] GUI Interface
- [ ] Sharing of tasks with other users 


This application was created using Java with OOP principles :D

```java
public class Main {
    public static void main(String[] args) {
        System.out.println("I love Java!");
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

SelwynAng committed:
Rename, Greet, Exit

SelwynAng committed:
Echo

SelwynAng committed:
Added adding and listing tasks functionality

SelwynAng committed:
Added functionality for marking tasks done

SelwynAng committed:
Added support for todos, deadlines, events

SelwynAng committed:
Added test using the I/O redirection technique

SelwynAng committed:
Added handling of exceptions

SelwynAng committed:
Added functionality for deleting tasks, more error handling

SelwynAng committed:
Added enums and refactored code

SelwynAng committed:
Updated createTodo method to handle empty args

SelwynAng committed:
Change indentation for Switch statments

SelwynAng committed:
Add save and retrive task functionality

SelwynAng committed:
Add Date and Time parsing functionality

SelwynAng committed:
Refactor code to extract out Storage, Parser, Ui, Command classes

SelwynAng committed:
Divide classes into packages

SelwynAng committed:
Change indention of switch statements

SelwynAng committed:
Build the project

SelwynAng committed:
Merge branch 'add-gradle-support' into branch-A-Gradle
Merge branch add-gradle support into branch-A-Gradle

SelwynAng committed:
Set up Gradle to run Duke and unit tests

SelwynAng committed:
Edit String output for certain methods

SelwynAng committed:
Add getIsDone method to Task file

SelwynAng committed:
Add unit testing for Parser and Task classes

SelwynAng committed:
Change Javadocs for greet method in Ui.java

SelwynAng committed:
Tweak code to comply with coding standard

SelwynAng committed:
Add Find Command functionality

SelwynAng committed:
Merge branch 'branch-Level-7'
Merge branch 'branch-Level-7' into master to add save and retrieve tasks functionality

SelwynAng committed:
Merge branch 'branch-Level-8'
Merge branch-Level-8 into master to add date and timing parsing functionality

SelwynAng committed:
Merge branch 'branch-A-MoreOOP'
Merge branch-A-MoreOOP into master to make code more OOP

SelwynAng committed:
Merge branch 'branch-A-Packages'
Merge branch-A-Packages into master to add packages

SelwynAng committed:
Merge branch 'add-gradle-support'
Add gradle support

SelwynAng committed:
Merge branch 'branch-A-Gradle'
Add gradle functionality

SelwynAng committed:
Merge branch 'branch-A-JUnit'
Add Java testing

SelwynAng committed:
Merge branch 'branch-A-JavaDoc'
Merge branch-A-JavaDoc into master to update Javadocs

SelwynAng committed:
Merge branch 'branch-A-CodingStandard'
Merge branch-A-CodingStandard into master to tweak code to comply with a coding standard

SelwynAng committed:
Merge branch 'branch-Level-9'
Merge branch-Level-9 into master to add find command functionality

SelwynAng committed:
Implement CheckStyle rules for all files

SelwynAng committed:
Merge branch 'branch-A-CheckStyle'
Merge branch-A-CheckStyle into master

lyuanww commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,84 @@
+        boolean exitProgram = false;
+        String userInput;
+
+        while (!exitProgram) {
```
Maybe you can use a variable that sounds more like a boolean e.g. `shouldExitProgram`?

lyuanww commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,273 @@
+     * @throws DukeException If the format is incorrect or the end date/time is invalid.
+     */
+    public Deadline createDeadline(String args) throws DukeException {
+        DateTimeFormatter dateTimeFormat = DateTimeFormatter.ofPattern("dd-MM-yyyy HHmm");
```
Perhaps you can make the format string `"dd-MM-yyyy HHmm"` a static variable?

lyuanww commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,273 @@
+     * @throws DukeException If the format is incorrect or the date/time is invalid.
+     */
+    public Event createEvent(String args) throws DukeException {
+        DateTimeFormatter dateTimeFormat = DateTimeFormatter.ofPattern("dd-MM-yyyy HHmm");
```
ditto

lyuanww commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,84 @@
+     * @param filePath The file path where task data is stored.
+     */
+    public Duke(String filePath) {
+        ui = new Ui(NAME);
+        parser = new Parser();
```
Maybe use `this` for the instance variables to reduce ambiguity?

lyuanww commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,84 @@
+        }
+
+        System.out.println(dirPath);
+        storage = new Storage(dirPath, dirAndFilePathArr[dirAndFilePathArr.length - 1]);
```
ditto

lyuanww commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,273 @@
+     * @throws DukeException If there is an issue with task creation or task type is unsupported.
+     */
+    public Task addTask(TaskType taskType, String args) throws DukeException {
+        Task newTask = null;
```
Perhaps the initialisation to null is unnecessary?

lyuanww commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,56 @@
+ * @author selwyn
+ */
+public class Parser {
+    /**
```
Maybe a blank line for better readability?

lyuanww reviewed (commented):
Overall, I like how all your classes are grouped into different packages. The abstraction of the Command class is commendable. 👍

SelwynAng committed:
Implement GUI for Duke application

SelwynAng committed:
Merge branch 'branch-Level-10'
Merge Level-10 branch into master

andytoh1 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,139 @@
+            throw new DukeException(e.getMessage());
+        }
+
+        return taskList;
```
This method is rather long, perhaps logic such as marking the task as done could be consolidated rather than done individually in each case.

andytoh1 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,112 @@
+     */
+    public static String printAddedTask(Task addedTask, int noOfTasksAfterAdding) {
+        String strToReturn = "";
+        strToReturn += "Got it. I've added this task: \n" + addedTask.toString() + "\n";
```
Since these are consecutive lines, perhaps they could be combined into a single line.

andytoh1 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,112 @@
+     * @param noOfTasksAfterAdding The number of tasks in the list after adding the new task.
+     * @return A message indicating the successful addition of the task.
+     */
+    public static String printAddedTask(Task addedTask, int noOfTasksAfterAdding) {
```
Perhaps "noOfTasksAfterAdding" could be named "numOfTasksAfterAdding" for clarity and consistency.

andytoh1 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,280 @@
+        } catch (DukeException e) {
+            throw new DukeException(e.getMessage());
+        }
+    }
```
As with several other instances, perhaps handling all the steps of parsing, validation, and error handling through calls to other methods will make these methods more adherent to the SLAP principle.

SelwynAng committed:
Improve UI of dialog box

SelwynAng committed:
Change build.gradle file to change javafx version

SelwynAng committed:
Improve dialog box appearance

SelwynAng committed:
Add assertions to files

Directory path in Duke and execute methods in Command classes may
contain null variables.

Null variables may cause the program to become illogical and crash.

Assertions are added to check if these relevant variables are null
at runtime and if they are, the program will exit.

Assertions help to ensure that the assumptions stated in the code
about the program will remain adhered to throughout runtime.

SelwynAng committed:
Merge pull request #1 from SelwynAng/branch-A-Assertions

Add assertions to files

SelwynAng committed:
Improve code quality

CodeQuality does not adhere to SALP principles

Adherance to SLAP principles is needed to ensure common levels of abstractions
in methods of the Command and Task related classes.

Long methods are broken down into more methods, with each method handling a
small portion of the logic.

This is to ensure that the methods are not overly complex and adhere to SLAP
principles.

SelwynAng committed:
Merge master to branch-A-CodeQuality

SelwynAng committed:
Merge pull request #2 from SelwynAng/branch-A-CodeQuality

Branch a code quality

SelwynAng committed:
Add Statistics Functionality

SelwynAng committed:
Merge branch 'branch-C-Stats'
Merge branch-C-Stats into master

SelwynAng committed:
Add CI workflows

SelwynAng committed:
Merge branch 'branch-A-CI'
Merge branch-A-CI into master

SelwynAng committed:
Improve coding standards

SelwynAng committed:
Improve the GUI

SelwynAng committed:
Merge branch 'branch-A-BetterGui'
Merge branch-A-BetterGui into master

SelwynAng committed:
Edit boolean variable to make it sound more boolean-like

SelwynAng committed:
Change statePercentOfTasksDone method to format a 0 percent correctly

SelwynAng committed:
Add Ui screenshot

SelwynAng committed:
Update README.md

SelwynAng committed:
Update README.md

SelwynAng committed:
Update README.md

SelwynAng committed:
Update README.md

SelwynAng committed:
Fix bug where percent of tasks is NaN when total number of tasks is zero

SelwynAng committed:
Update README.md

SelwynAng committed:
Edit gradle file to remove extra javafxplugin lines
